cluster_mapping.go

```diff
@@ -0,0 +1,202 @@
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package storage
+
+import (
+	"sync/atomic"
+	"time"
+
+	"github.com/facebookgo/clock"
+	"github.com/m3db/m3x/log"
+	"github.com/m3db/m3x/time"
+)
+
+// A ClusterMappingProvider provides cluster mapping rules
+type ClusterMappingProvider interface {
+	// Mappings returns all of the cluster mappings
+	Mappings() (ShardClusterMappings, error)
+}
+
+// ShardClusterMappings is the set of cluster mappings for all shards
+type ShardClusterMappings interface {
+	// MappingsFor returns the currently in-effect mappings for the given shard
+	// and retention policy.
+	MappingsFor(shard uint32, policy RetentionPolicy) []ClusterMapping
+}
+
+// A ClusterMapping defines which cluster holds the datapoints within a given timeframe
+type ClusterMapping interface {
+	// Cluster is the cluster that is targeted by this mapping
+	Cluster() string
+	SetCluster(c string) ClusterMapping
+
+	// CutoffTime defines the time that reads from FromCluster should stop.  Will
+	// inherently fall after the WriteCutoverTime, to account for configuration
+	// changes not arriving at all writers simultaneously
+	CutoffTime() time.Time
+	SetCutoffTime(t time.Time) ClusterMapping
+
+	// ReadCutoverTime defines the time that reads to ToCluster should begin
+	ReadCutoverTime() time.Time
+	SetReadCutoverTime(t time.Time) ClusterMapping
+
+	// WriteCutoverTime defines the time that writes to ToCluster should begin.  Will
+	// inherently fall after ReadCutoverTime, to account for configuration changes
+	// reaching writers ahead of readers.
+	WriteCutoverTime() time.Time
```

Comment:
If there are multiple cluster mappings whose cutover times are earlier than the time of the write, does the writer always pick the one with the latest cutover time so that a datapoint only goes to one cluster?